cranelift/codegen/src/egraph/node.rs

```diff
@@ -0,0 +1,382 @@
+                    types: ref other_types,
+                },
+            ) => {
+                *op == *other_op
+                    && self.ids_eq(args, other_args, uf)
+                    && types.as_slice(&self.types) == other_types.as_slice(&self.types)
+            }
+            (
+                &Node::Inst { inst, ref args, .. },
+                &Node::Inst {
+                    inst: other_inst,
+                    args: ref other_args,
+                    ..
+                },
+            ) => inst == other_inst && self.ids_eq(args, other_args, uf),
+            (
+                &Node::Load {
+                    ref op,
+                    ty,
+                    addr,
+                    mem_state,
+                    ..
+                },
+                &Node::Load {
+                    op: ref other_op,
+                    ty: other_ty,
+                    addr: other_addr,
+                    mem_state: other_mem_state,
+                    // Explicitly exclude: `inst` and `srcloc`. We
+                    // want loads to merge if identical in
+                    // opcode/offset, address expression, and last
+                    // store (this does implicit
+                    // redundant-load-elimination.)
+                    ..
+                },
+            ) => {
+                op == other_op
+                    && ty == other_ty
+                    && self.id_eq(addr, other_addr, uf)
+                    && mem_state == other_mem_state
+            }
+            _ => false,
+        }
+    }
+}
+
+impl CtxHash<Node> for NodeCtx {
+    fn ctx_hash(&self, value: &Node, uf: &mut UnionFind) -> u64 {
+        let mut state = crate::fx::FxHasher::default();
+        std::mem::discriminant(value).hash(&mut state);
+        match value {
+            &Node::Param {
+                block,
+                index,
+                ty: _,
+                loop_level: _,
+            } => {
+                block.hash(&mut state);
+                index.hash(&mut state);
+            }
+            &Node::Result {
+                value,
+                result,
+                ty: _,
+            } => {
+                self.hash_id(value, &mut state, uf);
+                result.hash(&mut state);
+            }
+            &Node::Pure {
+                ref op,
+                ref args,
+                types: _,
+            } => {
+                op.hash(&mut state);
+                self.hash_ids(args, &mut state, uf);
+                // Don't hash `types`: it requires an indirection
+                // (hence cache misses), and result type *should* be
+                // fully determined by op and args.
+            }
+            &Node::Inst { inst, ref args, .. } => {
+                inst.hash(&mut state);
+                self.hash_ids(args, &mut state, uf);
+            }
+            &Node::Load {
+                ref op,
+                ty,
```

Comment:
I appreciated the comment above that `types` are fully determined by `op` and `args`. I think it'd be nice if there's a corresponding comment here explaining why that isn't also true for Load ops. I assume it's that the type is the only thing indicating the width of the load; multiple loads with the same address can have different widths and must not be deduplicated.

Reply:
Done!